### PR TITLE
Add DeepSeek-V4 Flash Vision EXP support

### LINKED ISSUE #2135: Add DeepSeek-V4 Flash Vision EXP support
## Description

Add multimodal support for [`deepseek-ai/DeepSeek-V4-Flash-Vision-Exp`](https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-Vision-Exp).

`mlx-vlm` already supports the text-only DeepSeek-V4 architecture, but this checkpoint adds a vision tower, a spatial aligner, image-specific attention rules, vision-aware MoE routing, and checkpoint-local DSpark stages. Loading the checkpoint as the existing text-only model leaves its vision and routing parameters unused and cannot produce correct multimodal outputs.

This is a continued-trained multimodal checkpoint, so its vision parameters cannot safely be combined with an older converted DeepSeek-V4 Flash language checkpoint. The complete target checkpoint and its own DSpark weights need to be converted together.

## Required support

- Keep existing DeepSeek-V4 checkpoints text-only when `vision_n_layers == 0`.
- Add the published 32-layer vision transformer and 3x3 downsampling aligner.
- Implement the official image resize, patch, N-layout, type, permutation, and out-of-vocabulary sentinel-token processing.
- Accept OpenAI-style `image`, `image_url`, and `input_image` content blocks, including multiple images and mixed text/image batches.
- Apply bidirectional visibility within image spans while retaining causal behavior for surrounding text and supporting compression ratios 0, 4, and 128.
- Route image positions with `bias_vl`, including score-based routing in the hash-routed layers where image sentinel IDs cannot index `tid2eid`.
- Convert the published mixed FP8/FP4 checkpoint incrementally without loading all 168 GB of source weights at once, while preserving BF16 vision weights.
- Extract the checkpoint's three native DSpark stages into a compatible standalone drafter.
- Reuse packed vision features safely during generation and avoid splitting an image span across chunked prefill.

## Expected behavior

After conversion, the model should support text-only and multimodal generation, single and multiple images, mixed batches, repeated decode, continuous batching, cancellation, and its checkpoint-local DSpark drafter. Existing DeepSeek-V4 text-only checkpoints should continue to construct and run without vision modules.

Processor arrays, aligned vision features, and first-token logits should match the official implementation on a fixed image and prompt within the tolerances required by the published mixed-precision weights.

### PR DESCRIPTION
**Pull request by byoan**
Fixes #2135

## Summary

This PR adds experimental end-to-end support for [`deepseek-ai/DeepSeek-V4-Flash-Vision-Exp`](https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-Vision-Exp), including preprocessing, model execution, checkpoint conversion, batching, vision-feature caching, and the checkpoint-local DSpark drafter.

The existing DeepSeek-V4 implementation remains the language backbone. Vision is opt-in through `vision_n_layers`, so current text-only V4 checkpoints retain their existing construction and behavior.

## Changes

- Extend `deepseek_v4.ModelConfig` with the published vision configuration.
- Add the 32-layer bidirectional ViT with flattened 3x14x14 patch projection, two-dimensional rotary embeddings, RMSNorm/SwiGLU blocks, and the 3x3 spatial downsampling aligner.
- Implement the official resize constraints, RGB normalization, patch packing, N-layout construction, image types, permutations, and sentinel expansion.
- Preserve ordered OpenAI-style image content blocks and support packed, variable-size images across single-image, multi-image, and mixed batches.
- Embed safe vocabulary IDs before replacing the five out-of-vocabulary image sentinels with aligned features and learned start/end/newline/padding embeddings.
- Add image-span visibility to the DeepSeek-V4 local/compressed attention paths for compression ratios 0, 4, and 128, while keeping surrounding text causal.
- Make MoE routing vision-aware with `bias_vl`; hash-routed image positions use score-based selection rather than indexing `tid2eid` with sentinel IDs.
- Keep image prefill atomic, reject image sentinels during decode, and pass compatible target hidden states to text-only DSpark decode.
- Add a dedicated streaming converter for the checkpoint's mixed FP8/FP4 layout. It normalizes upstream namespaces, stacks routed experts, converts packed expert and quantization metadata to MLX layouts, and processes dependent shard groups incrementally without replacing the shared `mlx_vlm.convert` flow.
- Detect and extract all three native DSpark stages into separate indexed MLX shards, including the Markov and confidence heads.
- Cache packed per-image vision features and preserve image metadata through continuous-batching and cache-reuse paths.
- Document conversion and provide an exporter for the opt-in official-reference parity tests.

## Conversion

```bash
python -m mlx_vlm.models.deepseek_v4.convert \
  --hf-path deepseek-ai/DeepSeek-V4-Flash-Vision-Exp \
  --mlx-path DeepSeek-V4-Flash-Vision-Exp-MLX
```

Pass `--mtp` to extract the matching DSpark drafter beside the converted target. This dedicated converter preserves the published mixed FP8/FP4 layout, with the vision tower and aligner remaining BF16.
(The dedicated command is voluntary to avoid putting too much model specific logic in the classic convert command)

## Compatibility

- Text-only DeepSeek-V4 configs continue to default to `vision_n_layers=0` and do not construct vision modules.
- The generic conversion path is unchanged for other model families and DeepSeek-V4 text checkpoints.
- The DSpark drafter must be extracted from this checkpoint; an older V4 Flash drafter is not compatible with the continued-trained vision target.

## Validation

Focused deterministic coverage includes:

- processor tokens, resize constraints, N-layout, patch packing, and permutations;
- portrait, landscape, extreme-aspect-ratio, multiple-image, and mixed-batch inputs;
- vision/aligner shapes and image-embedding insertion;
- image-span attention and vision-aware routing across compression ratios 0, 4, and 128;
- repeated decode, chunked-prefill boundaries, continuous batching, cancellation, APC sentinel handling, and vision-feature-cache reuse;
- mixed-checkpoint conversion, three-stage DSpark extraction, and sanitizer idempotency.

Current focused test runs:

```text
102 passed, 3 skipped, 6 subtests passed
```

The three skipped tests are the opt-in real-checkpoint gates in `test_deepseek_v4_reference.py`. They compare processor arrays, aligned vision features against a fixture produced by `export_deepseek_v4_reference.py` with DeepSeek's official runtime when `DEEPSEEK_V4_VISION_MLX_PATH` and `DEEPSEEK_V4_VISION_FIXTURE` are set.

**See `docs/deepseek-v4-vision.md` for details on how to generate and run them.**

Results below:

| Parity stage | Result | Max absolute error | Mean absolute error | Notes |
|---|---:|---:|---:|---|
| Token IDs | Pass | `0` | `0` | Exact match |
| Image layout arrays | Pass | `0` | `0` | Grid, sample indices, offsets, types, type offsets, and permutations matched exactly |
| Normalized patches at BF16 vision-input stage | Pass | `0.0` | `0.0` | Compared after the MLX FP32 processor output was cast to BF16, matching the official processor stage |
| Patch embedding | Pass | `0.00390625` | `1.8350292e-08` | BF16-level numerical difference |
| Vision block 0 | Pass | `0.015625` | `0.0003437286` | BF16-level numerical difference |
| Final vision normalization | Informational | `0.453125` | `0.0015188056` | Sparse pre-aligner outliers; final aligned features remain within the reference tolerance |
| Aligner before permutation | Pass | `0.08306885` | `0.001401203` | Compared with `atol=0.1`, `rtol=0.02` |
| Aligner after checkpoint permutation | Pass | `0.08306885` | `0.001401203` | 99th percentile: `0.01245117`; 99.99th percentile: `0.05042989` |
| First-token logits | Not measured | — | — | The official language path requires CUDA/TileLang and could not run on Apple Silicon; no `first_token_logits` were written to the reference fixture |

The full converted checkpoint has also been successfully run on an M3 Ultra with the following image:

```bash
python3 -m mlx_vlm.generate \
  --model ./DeepSeek-V4-Flash-Vision-Exp-MLX \
  --image ./deepseek.jpg \
  --prompt "Describe this image in detail." \
  --max-tokens 256 \
  --temperature 0

The image displays the logo for **DeepSeek**, an artificial intelligence research company.

The logo consists of two main elements on a plain white background:

1.  **The Icon (Left):** A stylized, abstract illustration of a whale or dolphin in a vibrant blue color. The creature is depicted in a curved, circular shape, with its body arching upwards and its tail fin pointing towards the top right. The design uses negative space to create the appearance of a white belly and a small, simple eye. The overall shape is smooth and modern.
2.  **The Text (Right):** The word "deepseek" is written in a clean, lowercase, sans-serif font. The text is in the exact same shade of blue as the whale icon. The letters are bold and have a slightly rounded, modern feel.

The logo is set against a plain white background, giving it a clean, minimalist, and professional appearance. The blue color used is a vibrant, medium-bright shade, often associated with technology, intelligence, and trust.
```

deepseek

And with the DFlash drafter:

```bash
python3 -m mlx_vlm.generate \
  --model ./DeepSeek-V4-Flash-Vision-Exp-MLX \
  --draft-model ./DeepSeek-V4-Flash-Vision-Exp-DSpark-MLX \
  --image ./deepseek.jpg \
  --prompt "Describe this image in detail." \
  --max-tokens 256 \
  --temperature 0
Loading drafter (auto): ./DeepSeek-V4-Flash-Vision-Exp-DSpark-MLX
  → auto-detected --draft-kind='dflash'.
The image displays the logo for **DeepSeek**, an artificial intelligence research company.

The logo consists of two main elements on a plain white background:

1.  **The Icon (Left):** A stylized, abstract illustration of a whale or dolphin in a vibrant blue color. The creature is depicted in a curved, circular shape, with its tail fin pointing upwards and to the right. It features a white cutout for the eye and a white curved line along its belly, giving it a sleek, modern appearance.
2.  **The Text (Right):** The word "deepseek" is written in a clean, lowercase, sans-serif font. The text is in the exact same shade of blue as the whale icon.

The overall design is minimalist and modern, using a monochromatic blue color scheme against a plain white background.
Speculative decoding: 2.88 accepted tokens/round (1.88 accepted drafts/round, 37.6% of drafted, avg draft 5.00) over 59 rounds
```

Let me know if you need any other tests or validation

### CODE REVIEW

File: mlx_vlm/speculative/drafters/deepseek_v4_dspark/README.md [hunk 1]
```diff
@@ -1,7 +1,8 @@
 # DeepSeek-V4 DSpark drafter
 
 `deepseek_v4_dspark` is the native DSpark speculative head shipped inside
-`deepseek-ai/DeepSeek-V4-Flash-0731`. It is a **DeepSeek-V4-backbone variant of
```
Comment (Lazarus-931) [inline mlx_vlm/speculative/drafters/deepseek_v4_dspark/README.md:4]: i'm curious why you'r removing 0731 in place of this model

**Comment by Lazarus-931** [comment]
hi @byoan, thanks for the pr!

do you have a baseline vs mtp comparison?

**Comment by byoan** [comment]
@Lazarus-931 Yes, here it is:

DeepSeek-V4-Flash-Vision-Exp MLX target + extracted DSpark drafter, Apple M3 Ultra (512 GB), batch 1, greedy decoding, 128 generated tokens. Decode throughput excludes prefill and the first token. Every row produced an exact 128/128-token match against baseline.

| ctx | baseline tok/s | DSpark tok/s | speedup |
| ---: | ---: | ---: | ---: |
| 512 | 28.80 | 75.63 | 2.63x |
| 2048 | 26.05 | 40.75 | 1.56x |
| 4096 | 25.50 | 42.48 | 1.67x |
| 8192 | 25.35 | 39.81 | 1.57x |
| 16384 | 25.31 | 39.27 | 1.55x |

Additional DSpark stats:

Context | Accepted tokens/round | Draft acceptance | Token parity
-- | -- | -- | --
512 | 5.77 | 95.5% | 128/128
2,048 | 5.33 | 88.9% | 128/128
4,096 | 5.57 | 93.8% | 128/128
8,192 | 5.57 | 92.1% | 128/128
16,384 | 5.67 | 94.6% | 128/128

_(All rows in the table achieved exact 128/128-token parity on the deterministic continuation benchmark. As also observed with quantized GLM-5.3 DFlash2 targets, an additional open-ended prompt produced a single early baseline/speculative divergence)_

Let me know if you need anything else, or would like a specific test/bench to be run :)

**Review by Lazarus-931 — APPROVED** [review]

File: mlx_vlm/speculative/drafters/deepseek_v4_dspark/README.md (hunk 1 above)
Comment (byoan) [reply mlx_vlm/speculative/drafters/deepseek_v4_dspark/README.md:4]: Just wanted to mention the latest model available for that family, but I can open a follow-up to mention both if you'd prefer

Comment (Lazarus-931) [reply mlx_vlm/speculative/drafters/deepseek_v4_dspark/README.md:4]: no need, thanks :)
